data/cogctl.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH cogctl 1 "Jan 25, 2020"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
cogctl \- tool to control a running Cog instance
.SH SYNOPSIS
.B cogctl
.RI [ options ]
.RI < command >
.RI [ command\ options ]
.SH DESCRIPTION
\fBcogctl\fP is a tool to control a running Cog instance.

.SH OPTIONS
.TP
.B \-h,\ \-\-help
Show help options
.TP
.B \-A,\ \-\-appid=ID
Application identifier of the Cog instance to control
.TP
.B \-o,\ \-\-object\-path=PATH
Object path implementing the org.gtk.Actions interface
.TP
.B \-y,\ \-\-system
Use the system bus instead of the session bus

.SH COMMANDS
.TP
.B help [command]
Show a list of commands. If followed by a command name then it shows
additional information about that command.
.TP
.B appid
Display application ID being remotely controlled
.TP
.B objpath
Display the D-Bus object path being used
.TP
.B open <URL>
Open a URL
.TP
.B previous
Navigate backward in the page view history
.TP
.B next
Navigate forward in the page view history
.TP
.B ping
Check whether Cog is running
.TP
.B quit
Exit the application
.TP
.B reload
Reload the current page

.SH SEE ALSO
.BR cog (1)

.SH AUTHOR
This manual page was written by Alberto Garcia <berto@igalia.com>


NEWS
==========================
0.4.0 - September 20, 2019
==========================

- Avoid a deprecation warning when building with GLib 2.62.

===========================
0.3.91 - September 16, 2019
===========================

- HiDPI support. The "cog --device-scale" option can be used to set the
  initial scale factor.
- Using "cog --version" now also prints the WebKit version and (if available
  at build time), Git repository information.
- The libcogcore library now has proper libtool versioning.
- fdo: Fix axis scrolling.
- New "experimental-drm" platform plug-in which uses KMS+DRM (Kernel Mode
  Setting, Direct Rendering Manager) for direct display on framebuffers,
  and libinput for input event processing.

=====================
0.3.1 - June 28, 2019
=====================

- Fix installed cogcore.pc file to use the correct wpe-webkit-1.0 module as
  dependency.
- fdo: Use wayland-scanner and wayland-protocols to generate Wayland protocol
  code, instead of shipping pregenerated copies.
- fdo: Improved debug logging.

======================
0.3.0 - March 28, 2018
======================

- Support configuring the background color of the Web view.
- Add an option to control permission requests.
- The output of "cog --help-websettings" now is sorted alphabetically by
  option name.
- Properly handle URIs in the command line in some corner cases, for example
  when they include passwords or fragment identifiers.
- fdo: Avoid dereferencing a NULL view backend by creating the Wayland event
  source always after a valid backend is instantiated
- fdo: Set the opaque region of Wayland surfaces when fullscreened, which
  allows the compositor to optimize redraws.

=========================
0.2.0 - December 13, 2018
=========================

- fdo: Fix potential screen tearing by releasing each EGLImage on the
  release callback for its wl_buffer.
- fdo: Explicitly set the backend library name.
- Fix remotely controlling Cog over D-Bus.

=========================
0.1.2 - December 13, 2018
=========================

- Avoid error "xdg_surface not configured".
- Fix possible usage of a NULL Wayland EGL window object.

========================
0.1.1 - December 4, 2018
========================

- fdo: Support maximizing the Wayland surface by setting the
  COG_PLATFORM_FDO_VIEW_MAXIMIZE environment variable.
- Make it possible to build against GLib 2.58 and newer.
- Workaround WebKitGTK+ bug #150303.
- Properly set the application identifier in the generated D-Bus
  policy configuration file.

========================
0.1.1 - December 4, 2018
========================

- Initial release.


cogctl.c
/*
 * cogctl.c
 * Copyright (C) 2018 Adrian Perez <aperez@igalia.com>
 *
 * Distributed under terms of the MIT license.
 */

#define COG_INSIDE_COG__ 1

#include "cog-config.h"
#include "core/cog-utils.h"

#include <gio/gio.h>
#include <stdlib.h>
#include <string.h>

#ifndef COG_DEFAULT_APPID
#  if COG_USE_WEBKITGTK
#    define COG_DEFAULT_APPID "com.igalia.CogGtk"
#  else
#    define COG_DEFAULT_APPID "com.igalia.Cog"
#  endif
#endif

#define GTK_ACTIONS_ACTIVATE "org.gtk.Actions", "Activate"
#define FDO_DBUS_PEER_PING   "org.freedesktop.DBus.Peer", "Ping"


static struct {
    char    *appid;
    char    *objpath;
    gboolean system_bus;
} s_options = {
    .appid = COG_DEFAULT_APPID,
};


static GOptionEntry s_cli_options[] = {
    { "appid", 'A', 0, G_OPTION_ARG_STRING, &s_options.appid,
        "Application identifier of the Cog instance to control",
        "ID" },
    { "object-path", 'o', 0, G_OPTION_ARG_STRING, &s_options.objpath,
        "Object path implementing the org.gtk.Actions interface",
        "PATH" },
    { "system", 'y', 0, G_OPTION_ARG_NONE, &s_options.system_bus,
        "Use the system bus instead of the session bus",
        NULL },
    { NULL, }
};


static gboolean
call_method (const char *iface,
             const char *method,
             GVariant   *params,
             GError    **error)
{
    const GBusType bus_type =
        s_options.system_bus ? G_BUS_TYPE_SYSTEM : G_BUS_TYPE_SESSION;
    g_autoptr(GDBusConnection) conn = g_bus_get_sync (bus_type, NULL, error);
    if (!error)
        return FALSE;

    g_autoptr(GVariant) result =
        g_dbus_connection_call_sync (conn,
                                     s_options.appid,
                                     s_options.objpath,
                                     iface,
                                     method,
                                     params,
                                     NULL,
                                     G_DBUS_CALL_FLAGS_NO_AUTO_START,
                                     -1,
                                     NULL,
                                     error);
    return !!result;
}


struct cmd {
    const char *name;
    const char *desc;
    int (*handler) (const char *name, const void *data, int argc, char **argv);
    const void *data;
};

static const struct cmd* cmd_find_by_name (const char *name);


static void
cmd_check_simple_help (const char *name, int needed_argc, int *argc, char ***argv)
{
    const char *space = strchr (name, ' ');
    if (!space) space = strchr (name, '\0');

    g_autofree char *cmd_name = g_strndup (name, space - name);

    g_autoptr(GOptionContext) option_context = g_option_context_new (name);
    g_option_context_set_description (option_context,
                                      cmd_find_by_name (cmd_name)->desc);
    g_option_context_add_main_entries (option_context,
                                       &((GOptionEntry) { NULL, }),
                                       NULL);

    g_autoptr(GError) error = NULL;
    if (!g_option_context_parse (option_context, argc, argv, &error) || *argc > (1 + needed_argc)) {
        g_printerr ("%s: %s\n", name, error ? error->message : "No arguments expected");
        exit (EXIT_FAILURE);
    }
}


static int
cmd_generic_alias (const char *name,
                   const void *data,
                   int         argc,
                   char      **argv)
{
    const struct cmd *cmd = cmd_find_by_name ((const char*) data);
    return (*cmd->handler) (cmd->name, cmd->data, argc, argv);
}

static int
cmd_generic_no_args (const char               *name,
                     G_GNUC_UNUSED const void *data,
                     int                       argc,
                     char                    **argv)
{
    cmd_check_simple_help (name, 0, &argc, &argv);

    GVariant *params = g_variant_new ("(sava{sv})", name, NULL, NULL);

    g_autoptr(GError) error = NULL;
    if (!call_method (GTK_ACTIONS_ACTIVATE, params, &error)) {
        g_printerr ("%s\n", error->message);
        return EXIT_FAILURE;
    }
    return EXIT_SUCCESS;
}

/* A generic command applid to views. Receives the name of the view as
 * single argument. */
static int
cmd_generic_view (const char *name,
                   const void *data,
                   int         argc,
                   char      **argv)
{
    cmd_check_simple_help (name, 1, &argc, &argv);

    g_autoptr(GVariantBuilder) param_name =
        g_variant_builder_new (G_VARIANT_TYPE ("av"));
    g_variant_builder_add (param_name, "v", g_variant_new_string (argv[1]));

    GVariant *params = g_variant_new ("(sava{sv})", name, param_name, NULL);
    g_autoptr(GError) error = NULL;
    if (!call_method (GTK_ACTIONS_ACTIVATE, params, &error)) {
        g_printerr ("%s\n", error->message);
        return EXIT_FAILURE;
    }
    return EXIT_SUCCESS;
}

static int
cmd_appid (const char               *name,
           G_GNUC_UNUSED const void *data,
           int                       argc,
           char                    **argv)
{
    cmd_check_simple_help (name, 0, &argc, &argv);
    g_print ("%s\n", s_options.appid);
    return EXIT_SUCCESS;
}


static int
cmd_objpath (const char               *name,
             G_GNUC_UNUSED const void *data,
             int                       argc,
             char                    **argv)
{
    cmd_check_simple_help (name, 0, &argc, &argv);
    g_print ("%s\n", s_options.objpath);
    return EXIT_SUCCESS;
}


static int
cmd_open (const char               *name,
          G_GNUC_UNUSED const void *data,
          int                       argc,
          char                    **argv)
{
    cmd_check_simple_help ("open URL", 1, &argc, &argv);

    g_autoptr(GError) error = NULL;
    g_autofree char *utf8_uri =
        cog_uri_guess_from_user_input (argv[1], TRUE, &error);
    if (!utf8_uri) {
        g_printerr ("%s\n", error->message);
        return EXIT_FAILURE;
    }

    g_autoptr(GVariantBuilder) param_uri =
        g_variant_builder_new (G_VARIANT_TYPE ("av"));
    g_variant_builder_add (param_uri, "v", g_variant_new_string (utf8_uri));
    GVariant *params = g_variant_new ("(sava{sv})", "open", param_uri, NULL);

    if (!call_method (GTK_ACTIONS_ACTIVATE, params, &error)) {
        g_printerr ("%s\n", error->message);
        return EXIT_FAILURE;
    }
    return EXIT_SUCCESS;
}


static int
cmd_ping (const char               *name,
          G_GNUC_UNUSED const void *data,
          int                       argc,
          char                    **argv)
{
    cmd_check_simple_help (name, 0, &argc, &argv);
    return call_method (FDO_DBUS_PEER_PING, NULL, NULL)
        ? EXIT_SUCCESS
        : EXIT_FAILURE;
}


static int
cmd_help (const char *name,
          const void *data,
          int         argc,
          char      **argv)
{
    g_autoptr(GOptionContext) option_context =
        g_option_context_new ("help [command]");
    g_option_context_add_main_entries (option_context,
                                       &((GOptionEntry) { NULL, }),
                                       NULL);

    g_autoptr(GError) error = NULL;
    if (!g_option_context_parse (option_context, &argc, &argv, &error)) {
        g_printerr ("%s: %s\n", argv[0], error->message);
        return EXIT_FAILURE;
    }

    if (argc == 1) {
        g_print ("Available commands:\n");
        const struct cmd *cmds = data;
        for (unsigned i = 0; cmds[i].name; i++) {
            if (cmds[i].desc) {
                g_print ("  %-10s %s\n", cmds[i].name, cmds[i].desc);
            }
        }
    } else if (argc == 2) {
        const struct cmd *cmd = cmd_find_by_name (argv[1]);
        if (!cmd) {
            g_printerr ("Unrecognized command: %s\n", argv[1]);
            return EXIT_FAILURE;
        }

        if (cmd->handler == cmd_generic_alias) {
            g_print ("'%s' is aliased to '%s'\n", argv[1],  (const char*) cmd->data);
        } else {
            char *cmd_argv[] = { (char*) cmd->name, "--help" };
            return (*cmd->handler) (cmd->name, cmd->data, G_N_ELEMENTS (cmd_argv), cmd_argv);
        }
    } else {
        g_printerr ("Invalid command line ussage\n");
        return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}


static const struct cmd*
cmd_find_by_name (const char *name)
{
    static const struct cmd cmdlist[] = {
        {
            .name = "appid",
            .desc = "Display application ID being remotely controlled",
            .handler = cmd_appid,
        },
        {
            .name = "objpath",
            .desc = "Display the D-Bus object path being used",
            .handler = cmd_objpath,
        },
        {
            .name = "help",
            .desc = "Obtain help about commands",
            .data = cmdlist,
            .handler = cmd_help,
        },
        {
            .name = "open",
            .desc = "Open an URL",
            .handler = cmd_open,
        },
        {
            .name = "previous",
            .desc = "Navigate backward in the page view history",
            .handler = cmd_generic_no_args,
        },
        {
            .name = "prev",
            .data = "previous",
            .handler = cmd_generic_alias,
        },
        {
            .name = "next",
            .desc = "Navigate forward in the page view history",
            .handler = cmd_generic_no_args,
        },
        {
            .name = "ping",
            .desc = "Check whether Cog is running",
            .handler = cmd_ping,
        },
        {
            .name = "quit",
            .desc = "Exit the application",
            .handler = cmd_generic_no_args,
        },
        {
            .name = "reload",
            .desc = "Reload the current page",
            .handler = cmd_generic_no_args,
        },
        {
            .name = "add",
            .desc = "Add a new view",
            .handler = cmd_generic_view,
        },
        {
            .name = "close",
            .desc = "Close a view",
            .handler = cmd_generic_view,
        },
        {
            .name = "present",
            .desc = "Present a view",
            .handler = cmd_generic_view,
        },
        {
            .name = NULL,
        },
    };

    for (unsigned i = 0; cmdlist[i].name; i++)
        if (strcmp (cmdlist[i].name, name) == 0)
            return &cmdlist[i];

    return NULL;
}


static const char s_extended_help_text[] =
    "For a list of commands use:\n"
    "  cogctl help\n"
    "\n"
    "Help on each command can be obtained with one of:\n"
    "  cogctl help <command>\n"
    "  cogctl <command> --help\n";


int
main (int argc, char **argv)
{
    g_autoptr(GOptionContext) option_context =
        g_option_context_new ("<subcommand> [SUBCOMMAND-OPTION?]");
    g_option_context_set_description (option_context, s_extended_help_text);
    g_option_context_add_main_entries (option_context, s_cli_options, NULL);
    g_option_context_set_strict_posix (option_context, TRUE);

    g_autoptr(GError) error = NULL;
    if (!g_option_context_parse (option_context, &argc, &argv, &error)) {
        g_printerr ("Command line error: %s\n", error->message);
        return EXIT_FAILURE;
    }

    if (!g_application_id_is_valid (s_options.appid)) {
        g_printerr ("Invalid application ID: %s\n", s_options.appid);
        return EXIT_FAILURE;
    }

    if (!s_options.objpath) {
        s_options.objpath = cog_appid_to_dbus_object_path (s_options.appid);
    }
    if (!g_variant_is_object_path (s_options.objpath)) {
        g_printerr ("Invalid D-Bus object path: %s\n", s_options.objpath);
        return EXIT_FAILURE;
    }

    if (argc < 2) {
        g_printerr ("Missing subcommand\n");
        return EXIT_FAILURE;
    }

    const struct cmd *cmd = cmd_find_by_name (argv[1]);
    if (!cmd) {
        g_printerr ("Unrecognized command: %s\n", argv[1]);
        return EXIT_FAILURE;
    }

    return (*cmd->handler) (cmd->name, cmd->data, argc - 1, argv + 1);
}
